compiler/luci/service/src/Nodes/CircleFullyConnected.cpp

```diff
@@ -36,4 +40,76 @@ luci::CircleNode *CloneNodeLet<CN::DEF>::visit(const luci::CircleFullyConnected
   return cloned;
 }
 
+namespace sinf
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleFullyConnected *node)
+{
+  auto input_shape = circle_shape(loco::must_cast<CircleNode *>(node->input()));
+  auto weights_shape = circle_shape(loco::must_cast<CircleNode *>(node->weights()));
+
+  loco::TensorShape out_shape;
+
+  // NOTE Some recipes in some repositories are using rank 4 input for FullyConnected.
+  //      Until they are all fixed, disable following assert.
+  // TODO Enable following assert after related fixes are applied
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L194
+  // LUCI_ASSERT(input_shape.rank() == 2 || input_shape.rank() == 3,
+  //             "Input rank of FullyConnected should be 2 or 3");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L225
+  LUCI_ASSERT(weights_shape.rank() == 2, "Weights of FullyConnected should be 2");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L353-L367
+  if (node->keep_num_dims())
+  {
+    // [1,10,?,20] X [15,20] -> [1,10,?,15] (weights_shape =[15,20])
+    out_shape.rank(input_shape.rank());
+    for (uint32_t i = 0; i < input_shape.rank(); ++i)
+      out_shape.dim(i) = input_shape.dim(i);
+    out_shape.dim(out_shape.rank() - 1) = weights_shape.dim(0);
+  }
+  else
+  {
+    bool is_dynamic_shape = false;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      if (not input_shape.dim(i).known())
+      {
+        is_dynamic_shape = true;
+        break;
+      }
+    }
+
+    // Originally, BatchSize = input_size / weights_shape(0).value()
+    // The output BatchSize is determined by the input_size,
+    // which is calculated by multiplying dimensions up to rank()-2.
+
+    uint32_t batch_size = 1;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      batch_size *= input_shape.dim(i).value();
+    }
+
+    // input_shape = [batch_size, ..., weights_shape.dim(1).value()]
```

Comment:
Why is `input_shape` consist `weights_shape` ?

Reply:
> Why is `input_shape` consist `weights_shape` ?

I mean, input_shape's last dimension's value and weights_shape.dim(1).value() are same.

Reply:
`input_shape = [...]` reads as `input_shape` is made of [...]` which from my understanding, 
which doesn't make sense here.
if you want to express `input_shape's last dimension's value and weights_shape.dim(1).value() are the same.`, plz don't make it as `A = B` form.

Reply:
how about this expression?
```suggestion
    // input_shape = [batch_size, ..., n_inputs]
```

Reply:
where did `n_inputs` come from?
plz try to express how FC shape inference is calculate in short words,
equations can be interpreted in wrong direction from the author.

Reply:
I rechecked the FullyConnected operation.

As I understand it, the input shape of FullyConnected is like [batch_size, d0,d1,...,dN].
(In tensorflow,  'dN' refers to n_units.)
There is also a Weight_Tensor, which has the shape [out_units,input_dim(=n_units)].
The FullyConnected op computes the dot product between the inputs and weights.

1. if keep_num_dim is true,
The output shape's rank will be set as the input shape's rank.
the output shape is made of [batch_size,d0,d1,...,out_units]. ('out_units' means the number of output, or neurons).

3. if keep_num_dim is false,
The output shape's rank will be set 2.
It becomes [batch_size * d0*d1*...*dN-1, out_units].